File: haskell/glue/src/Glue/Lib/List/Remove.hs
module Glue.Lib.List.Remove where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

remove :: IR Eval
remove = NativeFunc removeImpl

removeImpl :: IR Eval -> Eval (IR Eval)
removeImpl itemIR = pure $ NativeFunc (removeFrom itemIR)

removeFrom :: IR Eval -> IR Eval -> Eval (IR Eval)
removeFrom item list = case list of
    List xs -> do
        let filtered = filter (/= item) xs
        pure $ List filtered
    _ -> throwError $ wrongArgumentType ["list"]

File: dart/glue/lib/src/lib/list/remove.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Remove function - removes all occurrences of an item from a list
/// Mirrors Haskell Glue.Lib.List.Remove.remove exactly
Ir remove = IrNativeFunc(removeImpl);

/// Remove function implementation
/// Mirrors Haskell Glue.Lib.List.Remove.removeImpl exactly
Eval<Ir> removeImpl(Ir itemIr) {
  return Eval.pure(IrNativeFunc(removeFrom(itemIr)));
}

Eval<Ir> Function(Ir) removeFrom(Ir item) {
  return (Ir list) => switch (list) {
    IrList(:final elements) => Eval.pure(
      IrList(elements.where((element) => element != item).toList()),
    ),
    _ => throwError(wrongArgumentType(['list'])),
  };
}
